src/connector/src/source/cdc/external/postgres.rs

```diff
@@ -265,6 +292,72 @@ impl PostgresExternalTableReader {
     }
 }
 
+mod scalar_adapter {
+    use pg_bigdecimal::PgNumeric;
+    use risingwave_common::types::ScalarRefImpl;
+    use tokio_postgres::types::{to_sql_checked, IsNull, Kind, ToSql, Type};
+
+    use crate::parser::EnumString;
+
+    #[derive(Debug)]
+    pub(super) enum ScalarAdapter<'a> {
+        Builtin(ScalarRefImpl<'a>),
+        Uuid(uuid::Uuid),
+        Numeric(PgNumeric),
+        Enum(EnumString),
+    }
+
+    impl ToSql for ScalarAdapter<'_> {
+        to_sql_checked!();
+
+        fn to_sql(
+            &self,
+            ty: &Type,
+            out: &mut bytes::BytesMut,
+        ) -> Result<IsNull, Box<dyn std::error::Error + Sync + Send>> {
+            match self {
+                ScalarAdapter::Builtin(v) => v.to_sql(ty, out),
+                ScalarAdapter::Uuid(v) => v.to_sql(ty, out),
+                ScalarAdapter::Numeric(v) => v.to_sql(ty, out),
+                ScalarAdapter::Enum(v) => v.to_sql(ty, out),
+            }
+        }
+
+        fn accepts(_ty: &Type) -> bool {
+            true
+        }
+    }
+
+    /// convert `ScalarRefImpl` to `ScalarAdapter` so that we can correctly encode to postgres value
+    pub(super) fn to_extra<'a>(
+        scalar: ScalarRefImpl<'a>,
+        ty: &Type,
+    ) -> super::ConnectorResult<ScalarAdapter<'a>> {
+        Ok(match (scalar, ty, ty.kind()) {
+            (ScalarRefImpl::Utf8(s), &Type::UUID, _) => ScalarAdapter::Uuid(s.parse().unwrap()),
```

Comment:
We shall either bubble the error up, or `unwrap` and simply return `ScalarAdapter<'a>`. Same for the function below.

Reply:
Refactor tracked in https://github.com/risingwavelabs/risingwave/issues/16580 and will submit a follow-up later